### DISCUSSION TITLE
Interactive Sequence digrams

### DISCUSSION THREAD

**Discussion by ThePlenkov** [likec4/likec4#218] [Ideas]
Hi! 

I liked so much this approach to design the system of any complexity and with any kinds of nodes.
I see so much potential in this tool. How about building sequence diagrams in a similar way with ability to split complicated process into seperate views with cross-navigation? Thanks!

**Comment by vedeja** [comment]
Structurizr’s dynamic views is what I currently use for that.

**Reply by ThePlenkov** [reply]
Do you have an example may be? Thank you

**Reply by vedeja** [reply]
Sure!

1. Go to https://structurizr.com/dsl
2. Insert the following DSL:
```
workspace {

    model {
        user = person "User"

        softwareSystem = softwareSystem "Software System" {
            webapp = container "Web Application" {
                user -> this "Uses"
            }
            
            db = container "Database" {
                webapp -> this "Reads from and writes to"
            }
            
            queue = container "Message Queue" {
                webapp -> this "Writes messages to"
            }
            
            listener = container "Queue Listener" {
                queue -> this "Forwards messages to"
                this -> db "Stores data in"
            }
        }
    }

    views {
        
        dynamic softwareSystem {
            user -> webapp "Clicks the update thing button"
            webapp -> db "Stores Thing in"
            webapp -> queue "Sends ThingUpdated message to"
            queue -> listener "Picks up the ThingUpdated message"
            listener -> db "Stores something"
        
            autolayout
        }

        theme default
    }

}
```
3. Click Render
4. Hover over the bottom right of the diagram and use the step forward/backward buttons to step through the sequence

**Reply by vedeja** [reply]
This would be something I'd like to have in likec4 before migrating from Structurizr

**Reply by davydkov** [reply]
Earlier versions of likec4 already had "dynamic" views, but I was not satisfied with the results. I am not a fan of the ones from Structurizr either.

After gathering different opinions and thoughts, I concluded that we lack some tooling to demonstrate "data/control" flows. We need something that is a bit less detailed than UML sequences, but more elaborate than flow/state diagrams.

@vedeja What are your thoughts? When do you use Structurizr dynamic views?

**Reply by vedeja** [reply]
Typically to show long flows that pass through many/all layers of our application. It has really helped in explaning how our very distributed solution works. Here is an example:
![image](https://github.com/likec4/likec4/assets/15958697/03e89a4d-5545-4e9e-8e15-14f2bdc79e42)

**Reply by simonwfarrow** [reply]
+1 for this feature.

I am switching from structurizr to likec4 but this is a feature I will miss.
The dynamic view in structurizr DSL allows you to redefine a view to change the relationship description and adds sequencing.

In likec4 , could the "extend view" function do the same, i.e. extend an existing view to include/exclude certain relationships whilst adding a sequence number based on the order they are defined?

**Reply by davydkov** [reply]
> 
> The dynamic view in structurizr DSL allows you to redefine a view to change the relationship description and adds sequencing.
> 

@simonwfarrow What do you think?

https://docs.likec4.dev/dsl/dynamic-views/

Just released yesterday

**Reply by vedeja** [reply]
@davydkov I have only tested it briefly so far, but it looks beautiful. And I especially appreciate the self-call feature! The new animated arrows and the automatic zoom to the current interaction makes the presentation very intuitive.  I think the control buttons grow to wide as the window gets wider. Not sure what your take on keyboard shortcuts are but, being able to use the keyboard to move between steps could be useful.

**Reply by davydkov** [reply]
Hey @vedeja, thank you for the feedback.
And reminding about shortcuts - created #769

**Reply by simonwfarrow** [reply]
> > The dynamic view in structurizr DSL allows you to redefine a view to change the relationship description and adds sequencing.
> 
> @simonwfarrow What do you think?
> 
> https://docs.likec4.dev/dsl/dynamic-views/
> 
> Just released yesterday

Yes looks fantastic, very timely 😄 
One thought, I think it applies to all views not just dynamic, when I have multiple relationships going from the same source to target it only renders one of them. Is there a way to display all relationships?

**Reply by aelberg** [reply]
Agree with the others that the new dynamic view is amazing.  It seems that a relationship between elements in the model is not required (as it was with structurizr) as well which makes the dynamic view here much easier to get up and running.
I am using this feature to create a diagram for event sagas where I don't want to clutter my model with all the event relationships and so far it is working very well.

Thank you!

@simonwfarrow I was able to view multiple relationships from the same source to target in the dynamic diagram I created.

**Comment by davydkov** [comment]
> I have multiple relationships going from the same source to target it only renders one of them. Is there a way to display all relationships?

@simonwfarrow, it is not possible for regular views, as all the relationships between two elements (or relationships between their nested ones) are merged into a single one with `[...]`.
It indicates that there are more relationships than what is currently displayed, and the most relevant title cannot be determined.

The idea behind `[...]` was to reveal the details on hover/click, like from which relationships this merged one was derived. 
But honestly, this was never demanded in my practice, and thus I have not implemented it yet.

You can try the reverse relationship:

```zig
  a -> b

  // this one has reverse direction
  // but arrows are also reversed (from head -> tail, i.e. a -> b)
  b -> a { 
    style {
      head none
      tail normal
    }
  }
```

<img width="1201" alt="image" src="https://github.com/likec4/likec4/assets/824903/3a04c036-5763-4a41-baa1-d5b564818537">

**Reply by vedeja** [reply]
The idea of revealing all connections on hover/click on […] is a good one.

**Reply by gparmeggiani** [reply]
hello!

I opened this a while ago https://github.com/likec4/likec4/issues/663

still hoping it will be implemented :)
